### PR TITLE
Ability to Cache Token into an external cache: Database, MongoDB, Redis, Memcache, etc

### PR DESCRIPTION
**Pull request by matiasperrone**
### Main Goal

To allow implement any external cache solutions such as MongoDB, MySQL, Oracle, SQL Server, Redis, Memcache or any other easily, by extending or implementing the same methods as the original class.

### Issue related
#278 Ability to Cache Token to MongoDB or MySql

### Code example of an external cache

In this case we use MongoDB with a proprietary implementation, but ilustrates how easy is to implement an external cache:

```php
<?php

namespace MyNamespace\PayPal;

// Composer Loader
include_once 'composer/vendor/autoload.php';
use PayPal\Cache\AuthorizationCache as PayPal_AuthorizationCache;


/**
 * Special modification introduced by Matias Perrone
 * @author		Matias Perrone <matias@aluguest.com>
 * @uses		Mongo class
 */

class AuthorizationCache extends PayPal_AuthorizationCache
{
	public static $mongo = false;
	public static $colection = 'PayPalAuthorizationCache';

	public static function init()
	{
		if ( !self::$mongo )
    		self::$mongo = new MyNamespace\MyMongo();
	}

    /**
     * A pull method which would read the persisted data based on clientId.
     * If clientId is not provided, an array with all the tokens would be passed.
     *
     * @param array|null $config
     * @param string $clientId
     * @return mixed|null
     */
    public static function pull($config = null, $clientId = null)
    {
        // Return if not enabled
        if (!self::isEnabled($config)) { return null; }

    	self::init();

    	$tokens = null;
    	$find = array();
		$qty = 100;
		$page = 1;
		$total = null;
    	$fields = array('_id' => false);

		if ($clientId)
		{
			$find['clientId'] = $clientId;
			$tokens = self::$mongo->getItem(self::$colection, $find, $fields);
			if (!$tokens)
				$tokens = null;
		}
		else
		{
			$tokens = self::$mongo->getAll(self::$colection, $qty, $page, $total, $find, $fields);
		}
        return $tokens;
    }

    /**
     * Persists the data into the "mongo" cache
     *
     * @param array|null $config
     * @param      $clientId
     * @param      $accessToken
     * @param      $tokenCreateTime
     * @param      $tokenExpiresIn
     * @throws \Exception
     */
    public static function push($config = null, $clientId, $accessToken, $tokenCreateTime, $tokenExpiresIn)
    {
        // Return if not enabled
        if (!self::isEnabled($config)) { return null; }

    	self::init();

        $data = array(
            'clientId' => $clientId,
            'accessTokenEncrypted' => $accessToken,
            'tokenCreateTime' => $tokenCreateTime,
            'tokenExpiresIn' => $tokenExpiresIn
        );
    	$expireAfterSeconds = $tokenExpiresIn;
    	$CreateIndexField = 'clientId';
    	$CreateIndexUnique = true;
        $ok = self::$mongo->saveData(self::$colection, $CreateIndexField, $data, $expireAfterSeconds, $CreateIndexUnique);
        if (!$ok)
        {
            throw new \Exception("Failed to write cache");
        };
    }
}
```

### PayPal API Usage:

Check *getOAuthTokenCredential* method to view how is implemented

```php
<?php
namespace MyNamespace\PayPal;

include_once APPPATH.'libraries/MyNamespace/PayPal/AuthorizationCache.php';
include_once APPPATH.'third_party/vendor/autoload.php';

use \MyNamespace\PayPal\AuthorizationCache;
use \PayPal\Auth\OAuthTokenCredential;
use \PayPal\Api\Amount;
use \PayPal\Api\Payer;
use \PayPal\Api\Payment;
use \PayPal\Api\PaymentExecution;
use \PayPal\Api\FundingInstrument;
use \PayPal\Api\RedirectUrls;
use \PayPal\Api\Transaction;
use \PayPal\Api\Item;
use \PayPal\Api\ItemList;
use \PayPal\Rest\ApiContext;

class PayPal
{
	protected $_paypal;
	public $_payment = false;
 	const collection = 'PayPal';

	public function __construct ($config)
	{
		$this->_paypal = $config;

	}

	public function &getOAuthTokenCredential()
	{
		$credential = new \PayPal\Auth\OAuthTokenCredential( $this->_paypal['paypal_clientid'], $this->_paypal['paypal_secret'] );
		$credential->AuthorizationCacheClass = '\\aluGuest\\PayPal\\AuthorizationCache';
		return $credential;
	}

	public function create(array &$params, $type)
	{

		// ### Payer
		// A resource representing a Payer that funds a payment
		// Use the List of `FundingInstrument` and the Payment Method
		// as 'credit_card'
		$payer = new Payer();
		$payer->setPaymentMethod("paypal");

		// ### Amount
		// Let's you specify a payment amount.
		$amount = new Amount();
		$amount->setCurrency("USD");
		$amount->setTotal($params['total']);

		// ### Item
		// Let's you specify a item info.
		$item = new Item();
		$item->setQuantity(1);
		$item->setCurrency("USD");
		$item->setPrice(intval($params['total']));
		$item->setName($params['description']);
		$item->setDescription($params['title']);
		$item->setSku($params['transaction']);

		// ### ItemList
		// Let's you specify a item info.
		$itemlist = new ItemList();
		$itemlist->addItem($item);

		// ### Transaction
		// A transaction defines the contract of a
		// payment - what is the payment for and who
		// is fulfilling it. Transaction is created with
		// a `Payee` and `Amount` types
		$transaction = new Transaction();
		$transaction->setItemList($itemlist);
		$transaction->setAmount($amount);
		$transaction->setCustom($params['transaction']);

		// ### Redirect urls
		// Set the urls that the buyer must be redirected to after
		// payment approval/ cancellation.
		$redirectUrls = new RedirectUrls();
		$redirectUrls->setReturnUrl($params['redirectOk']);
		$redirectUrls->setCancelUrl($params['redirectCancel']);

		// ### Payment
		// A Payment Resource; create one using
		// the above types and intent as 'sale'
		$this->_payment = new Payment();
		$this->_payment->setIntent($type);
		$this->_payment->setPayer($payer);
		$this->_payment->setRedirectUrls($redirectUrls);
		$this->_payment->setTransactions(array($transaction));


		// ### Api Context
		// Pass in a `ApiContext` object to authenticate
		// the call and to send a unique request id
		// (that ensures idempotency). The SDK generates
		// a request id if you do not pass one explicitly.
		$apiContext = new \PayPal\Rest\ApiContext($this->getOAuthTokenCredential(), $params['transaction']);
		$apiContext->setConfig( $this->_getConfig() );

		// ### Create Payment
		// Create a payment by posting to the APIService
		// using a valid apiContext
		// The return object contains the status and the
		// url to which the buyer must be redirected to
		// for payment approval
		try
		{
			$this->_payment->create($apiContext);
		}
		catch (\PPConnectionException $ex)
		{
			log_message('error', 'PAYPAL PPConnectionException: '.$ex->getMessage(). ' Data:');
			log_message('error', $ex->getData());
			return false;
		}
		catch (\PPConfigurationException $ex)
		{
			log_message('error', 'PAYPAL PPConfigurationException: '.$ex->getMessage(). ' Datos:');
			log_message('error', $ex->getData());
			return false;
		}
		catch (\Exception $ex)
		{
			log_message('error', 'PAYPAL EXCEPTION: '.$ex->getMessage());
			return false;
		}

		// ### Redirect buyer to paypal
		// Retrieve buyer approval url from the `payment` object.
		foreach($this->_payment->getLinks() as $link)
		{
			if($link->getRel() == 'approval_url')
			{
				$redirectUrl = $link->getHref();
			}
		}

		return $redirectUrl;
	}

	public function createAuthorize(array &$params)
	{
		return $this->create($params, 'authorize');
	}

	public function createSale(array &$params)
	{
		return $this->create($params, 'sale');
	}

	public function executePayment(array &$payment, $payerID, $paymentId)
	{
		if ( empty($paymentId) Or empty($payerID) )
			return false;

		// ### Api Context
		// Pass in a `ApiContext` object to authenticate
		// the call and to send a unique request id
		// (that ensures idempotency). The SDK generates
		// a request id if you do not pass one explicitly.
		$apiContext = new \PayPal\Rest\ApiContext($this->getOAuthTokenCredential());
		$apiContext->setConfig( $this->_getConfig() );

		// ### Payment
		// A Payment Resource; create one using
		// the above types and intent as 'sale'
		$this->_payment = Payment::get($paymentId, $apiContext);

		// ### Payer
		// A resource representing a Payer that funds a payment
		// Use the List of `FundingInstrument` and the Payment Method
		// as 'credit_card'
		$paymentEx = new PaymentExecution();
		$paymentEx->setPayerId($payerID);

		// ### Create Payment
		// Create a payment by posting to the APIService
		// using a valid apiContext
		// The return object contains the status and the
		// url to which the buyer must be redirected to
		// for payment approval
		try
		{
			$this->_payment = $this->_payment->execute($paymentEx, $apiContext);

			try
			{
				 $this->_payment = Payment::get($paymentId, $apiContext);
			}
			catch (\Exception $ex)
			{
				gi_log_message('error', 'PAYPAL EXCEPTION: luego de hacer Execute al buscar el pago:'.$ex->getMessage());
				return false;
			}
		}
		catch (\Exception $ex)
		{
			gi_log_message('error', 'PAYPAL EXCEPTION: al hacer Execute :'.$ex->getMessage());
			return false;
		}

		if ($this->_payment And $this->_payment->getState() == 'approved')
			return $paymentId;

		elseif ($this->_payment)
		{
			if ($this->_payment->getState() == 'pending')
				$payment['status'] = 'pending';
			else
				$payment['status'] = 'error';
		}

		return false;
	}

	private function _getConfig()
	{
		$config =
			array(
				'log.FileName' =>  APPPATH.'logs/paypal/paypal-'.date('Y-m-d').'.log',
				'log.LogLevel' => 'FINE',
				'cache.enabled' => true
			);

		if (array_key_exists('paypal_mode', $this->_paypal) And in_array($this->_paypal['paypal_mode'], array('sandbox', 'live')) )
			$config['mode'] = $this->_paypal['paypal_mode'];

		return $config;
	}
}
```

### CODE REVIEW

**Comment by jaypatel512** [comment]
Hey @matiasperrone ! 

Thank you so much for your pull request. Here are my few thoughts:

- I think it looks like the solution you are suggesting is not going to work as expected. The classname is a string, directly provided in the oauthcredential class, which would be impossible to change, if someone wish to use file based system instead of mongo, or redis or any other storage mechanism. 

- The mongo object you have created requires dependency to work. It will not work unless you download Mongo drivers. Also, the object initialization could be more involved then this. E.g. Passing Credentials, etc.

- We need a factory implementation or adapter to connect the AuthorizationCache object. Its implementation could be injected runtime. E.g. "RedisCache", "MongoCache", "DefaultCache" or similar. This way, anyone at runtime could either set up a config, or create an instance of these objects, and inject the data accordingly.

- Providing Factory object allows you to extend the class, and provide your way to initializing a mongo driver, instead of using a default constructor. 

I will still dig a little bit more into your pull request and get back to you on this. However, as you can see, getting this right is more important than getting it done. I wish to make sure we cover all scenarios before implementing this in our SDK.

**Comment by jaypatel512** [comment]
Btw, I am really impressed the way you have presented your pull request. Not many developers understand the importance of it.

**Comment by matiasperrone** [comment]
Jay,Thank you very much!

I mean to do it really easy, with no complexity.

It is true that it might lead to errors, but this approach requires minimum knowledge of the internal functions, and it took a few hours to find the function and the right place.

It's true that it might lead to errors and maybe it isn't the "best" way, but we (at aluGuest) found this easy to do and easy to maintain even with PayPal's SDK updates.

So that's why.

May I help you in some way?
We are using this in production, and works fine. :smiley:

**Comment by jaypatel512** [comment]
You definitely helped. This is definitely 80% there. All we need to do is, if you think about it, we could have similar modularity available for many other things, which makes it better if we come up with a standard that we follow across the entire SDK. This makes it easier for everyone to understand what is going on. 

This is really amazing. We really appreciate contributions from external developers. :smiley:

**Comment by yceruto** [comment]
some news for merge this PR?

**Comment by jaypatel512** [comment]
Hey @matiasperrone !

We had a similar design pattern implemented recently for Logging Interface. We allowed anyone to extend/inject their logging mechanism to it, as shown [here](https://github.com/paypal/PayPal-PHP-SDK/wiki/Custom-Logger). 

I think we would want to have a similar approach for this. I will see if I can take that task if possible soon, however, anyone is free to pick that up. It should be very easy to implement.

File: lib/PayPal/Auth/OAuthTokenCredential.php [hunk 1]
```diff
@@ -143,7 +149,7 @@ public function getAccessToken($config)
 
             // Case where we have an old unencrypted cache file
             if (!array_key_exists('accessTokenEncrypted', $token)) {
-                AuthorizationCache::push($config, $this->clientId, $this->encrypt($this->accessToken), $this->tokenCreateTime, $this->tokenExpiresIn);
+                $AuthorizationCacheClass::push($config, $this->clientId, $this->encrypt($this->accessToken), $this->tokenCreateTime, $this->tokenExpiresIn);
```
Comment (jaypatel512) [inline lib/PayPal/Auth/OAuthTokenCredential.php]: We need some checks to verify if the class is infact what we expect. Having an interface here would be perfect.

**Comment by jaypatel512** [comment]
Closing this PR based on inactivity. Please re-open if more people thinks its a good idea.

**Pull request closed without merging**